### PR TITLE
CES-9803 - Configure baremetal server's boot mode to UEFI or BIOS

### PR DESCRIPTION
**Pull request by digambar15**

### CODE REVIEW

**Comment by digambar15** [comment]
Below is test results - 

**Set UEFI Boot Mode -** 

[osp_admin@director pilot]$ python config_idrac.py 192.168.110.16 -n /home/osp_admin/instackenv.json

INFO:config_idrac:Resetting the iDRAC on 192.168.110.16
INFO:discover_nodes.dracclient.client:Waiting for the iDRAC to become not pingable
INFO:discover_nodes.dracclient.client:The iDRAC has become not pingable
INFO:discover_nodes.dracclient.client:Waiting for the iDRAC to become pingable
INFO:discover_nodes.dracclient.client:The iDRAC has become pingable
INFO:discover_nodes.dracclient.client:Waiting for the iDRAC to become ready
WARNING:dracclient.wsman:A ConnectionError error occurred while  communicating with 192.168.110.16, attempt 1 of 3
INFO:discover_nodes.dracclient.client:The iDRAC is ready
INFO:config_idrac:Clearing the job queue on 192.168.110.16
INFO:config_idrac:Configuring NIC NIC.Integrated.1-3-1 on 192.168.110.16 to PXE boot
INFO:config_idrac:Setting 192.168.110.16 to Uefi boot
WARNING:dracclient.resources.bios:Ignoring unchanged BIOS attributes: ['BootMode']
INFO:config_idrac:Configuring initial iDRAC settings on 192.168.110.16
INFO:config_idrac:Completed configuration of the iDRAC on 192.168.110.16

Result - 
	Current Boot Mode =  Uefi
	Pending Boot Mode =  None



**Set BIOS boot mode**

[osp_admin@director pilot]$ python config_idrac.py 192.168.110.16 -n /home/osp_admin/instackenv.json

INFO:config_idrac:Resetting the iDRAC on 192.168.110.16
INFO:discover_nodes.dracclient.client:Waiting for the iDRAC to become not pingable
INFO:discover_nodes.dracclient.client:The iDRAC has become not pingable
INFO:discover_nodes.dracclient.client:Waiting for the iDRAC to become pingable
INFO:discover_nodes.dracclient.client:The iDRAC has become pingable
INFO:discover_nodes.dracclient.client:Waiting for the iDRAC to become ready
WARNING:dracclient.wsman:A SSLError error occurred while  communicating with 192.168.110.16, attempt 1 of 3
INFO:discover_nodes.dracclient.client:The iDRAC is ready
INFO:config_idrac:Clearing the job queue on 192.168.110.16
INFO:config_idrac:Configuring NIC NIC.Integrated.1-3-1 on 192.168.110.16 to PXE boot
INFO:config_idrac:Setting 192.168.110.16 to Bios boot
INFO:config_idrac:Configuring initial iDRAC settings on 192.168.110.16
INFO:config_idrac:Rebooting 192.168.110.16 to apply configuration
INFO:config_idrac:Waiting for iDRAC configuration to complete on 192.168.110.16
INFO:config_idrac:Do not unplug 192.168.110.16
INFO:config_idrac:Completed configuration of the iDRAC on 192.168.110.16

Result -
		Current Boot Mode =  Bios
		Pending Boot Mode =  None

**Comment by digambar15** [comment]
Please review the code, all the changes are completed.

File: src/pilot/config_idrac.py [hunk 1]
```diff
@@ -339,12 +343,18 @@ def config_idrac(instack_lock,
...
+        boot_mode = boot_mode_helper.DRAC_BOOT_MODE_UEFI
+    else:
+        # Configure the system to BIOS boot mode
+        boot_mode = boot_mode_helper.DRAC_BOOT_MODE_BIOS
+
+    reboot_required_boot, boot_job_id = config_boot_mode(
+        drac_client, ip_service_tag, node, boot_mode)
+    reboot_required = reboot_required or reboot_required_boot
+    if boot_job_id:
+        job_ids.append(boot_job_id)
```
[the remark this answers is not in the document]
Comment (digambar15) [reply src/pilot/config_idrac.py]: Fixed

File: src/pilot/config_idrac.py [hunk 2]
```diff
@@ -355,6 +365,16 @@ def config_idrac(instack_lock,
     if idrac_job_id:
         job_ids.append(idrac_job_id)
 
+    if boot_mode == boot_mode_helper.DRAC_BOOT_MODE_UEFI and reboot_required:
```
[the remark this answers is not in the document]
Comment (digambar15) [reply src/pilot/config_idrac.py]: Done

File: src/pilot/config_idrac.py [hunk 3]
```diff
@@ -355,6 +365,16 @@ def config_idrac(instack_lock,
...
+        boot_job_ids = list()
+        boot_job_ids.append(idrac_job_id)
+        success = schedule_restart_job(boot_job_ids,
+                                       drac_client,
+                                       drac_ip,
+                                       drac_user,
+                                       password,
+                                       ip_service_tag)
```
[the remark this answers is not in the document]
Comment (digambar15) [reply src/pilot/config_idrac.py]: Removed the schedule_restart_job() function. Done

File: src/pilot/config_idrac.py [hunk 4]
```diff
@@ -339,12 +343,18 @@ def config_idrac(instack_lock,
...
 
-    # Configure the system to legacy boot
-    reboot_required_legacy, legacy_job_id = config_legacy_boot_mode(
-        drac_client, ip_service_tag, node)
-    reboot_required = reboot_required or reboot_required_legacy
-    if legacy_job_id:
-        job_ids.append(legacy_job_id)
+    if BootModeHelper.is_boot_order_flexibly_programmable(drac_client):
+        # Configure the system to UEFI boot mode
+        boot_mode = boot_mode_helper.DRAC_BOOT_MODE_UEFI
```
[the remark this answers is not in the document]
Comment (digambar15) [reply src/pilot/config_idrac.py]: Done

File: src/pilot/config_idrac.py [hunk 5]
```diff
@@ -35,6 +36,9 @@
 from discover_nodes.dracclient.client import DRACClient  # noqa
 from discover_nodes.dracclient.exceptions import NotFound  # noqa
 
+import boot_mode_helper
+from boot_mode_helper import BootModeHelper
```
[the remark this answers is not in the document]
Comment (digambar15) [reply src/pilot/config_idrac.py]: Done!

File: src/pilot/config_idrac.py [hunk 6]
```diff
@@ -153,10 +157,10 @@ def configure_nics_boot_settings(
     return reboot_required, job_ids, provisioning_mac
 
 
-def config_legacy_boot_mode(drac_client, ip_service_tag, node):
-    LOG.info("Setting {} to legacy BIOS boot".format(
-        ip_service_tag))
-    settings = {"BootMode": "Bios"}
+def config_boot_mode(drac_client, ip_service_tag, node, boot_mode):
+    LOG.info("Setting {} to {} boot".format(
+        ip_service_tag, boot_mode))
```
[the remark this answers is not in the document]
Comment (digambar15) [reply src/pilot/config_idrac.py]: Done!

File: src/pilot/config_idrac.py [hunk 7]
```diff
@@ -339,12 +343,18 @@ def config_idrac(instack_lock,
...
-    if legacy_job_id:
-        job_ids.append(legacy_job_id)
+    if BootModeHelper.is_boot_order_flexibly_programmable(drac_client):
+        # Configure the system to UEFI boot mode
+        boot_mode = boot_mode_helper.DRAC_BOOT_MODE_UEFI
+    else:
+        # Configure the system to BIOS boot mode
+        boot_mode = boot_mode_helper.DRAC_BOOT_MODE_BIOS
+
+    reboot_required_boot, boot_job_id = config_boot_mode(
```
[the remark this answers is not in the document]
Comment (digambar15) [reply src/pilot/config_idrac.py]: Done!

File: src/pilot/config_idrac.py [hunk 8]
```diff
@@ -15,11 +15,14 @@
 # limitations under the License.
 
 import argparse
+import boot_mode_helper
```
Comment (rpioso) [inline src/pilot/config_idrac.py]: Please move this line to the section below. This section is for Python standard library imports. For more information, see [PEP 8 -- Style Guide for Python Code](https://www.python.org/dev/peps/pep-0008/#imports).

File: src/pilot/config_idrac.py [hunk 9]
```diff
@@ -31,13 +34,11 @@
 discover_nodes_path = os.path.join(os.path.expanduser('~'),
                                    'pilot/discover_nodes')
 sys.path.append(discover_nodes_path)
-
-from discover_nodes.dracclient.client import DRACClient  # noqa
-from discover_nodes.dracclient.exceptions import NotFound  # noqa
+from discover_nodes.dracclient.client import DRACClient
+from discover_nodes.dracclient.exceptions import NotFound
```
Comment (rpioso) [inline src/pilot/config_idrac.py]: Why were these lines modified by removing the `# noqa` at the end? My editor now reports the following errors for these two (2) lines.

    E402 module level import not at top of file
    E402 module level import not at top of file

File: src/pilot/config_idrac.py [hunk 10]
```diff
@@ -31,13 +34,11 @@
 discover_nodes_path = os.path.join(os.path.expanduser('~'),
                                    'pilot/discover_nodes')
 sys.path.append(discover_nodes_path)
-
```
Comment (rpioso) [inline src/pilot/config_idrac.py]: Restore this blank line.

File: src/pilot/config_idrac.py [hunk 11]
```diff
@@ -31,13 +34,11 @@
...
-from discover_nodes.dracclient.client import DRACClient  # noqa
-from discover_nodes.dracclient.exceptions import NotFound  # noqa
+from discover_nodes.dracclient.client import DRACClient
+from discover_nodes.dracclient.exceptions import NotFound
 
 # Suppress InsecureRequestWarning: Unverified HTTPS request is being made
 requests.packages.urllib3.disable_warnings()
-
```
Comment (rpioso) [inline src/pilot/config_idrac.py]: Restore this blank line.

File: src/pilot/config_idrac.py [hunk 12]
```diff
@@ -153,10 +154,10 @@ def configure_nics_boot_settings(
     return reboot_required, job_ids, provisioning_mac
 
 
-def config_legacy_boot_mode(drac_client, ip_service_tag, node):
-    LOG.info("Setting {} to legacy BIOS boot".format(
-        ip_service_tag))
-    settings = {"BootMode": "Bios"}
+def config_boot_mode(drac_client, ip_service_tag, node, boot_mode):
+    LOG.info("Setting {} to {} boot".format(
+        ip_service_tag, (boot_mode.upper())))
```
Comment (rpioso) [inline src/pilot/config_idrac.py]: There appears to be an unnecessary pair of parentheses. Is not the following sufficient?

    ip_service_tag, boot_mode.upper()))

File: src/pilot/config_idrac.py [hunk 13]
```diff
@@ -302,7 +303,8 @@ def config_idrac(instack_lock,
     drac_user = node["pm_user"]
     drac_password = node["pm_password"]
     ironic_driver = node["pm_type"]
-
+    job_ids = []
```
Comment (rpioso) [inline src/pilot/config_idrac.py]: Delete this line. My editor warns, "Unused variable: job_ids". Note that line 336 below also defines `job_ids`, which is also before its first use.

File: src/pilot/config_idrac.py [hunk 14]
```diff
@@ -322,30 +324,57 @@ def config_idrac(instack_lock,
     # Clear out any pending jobs in the job queue and fix the condition where
     # there are no pending jobs, but the iDRAC thinks there are
     clear_job_queue(drac_client, ip_service_tag)
+    reboot_required = False
+
+    if BootModeHelper.is_boot_order_flexibly_programmable(drac_client):
+        # Set target_boot_mode to UEFI boot mode
```
Comment (rpioso) [inline src/pilot/config_idrac.py]: Again, this comment is unnecessary, because the code is self-explanatory. Please delete it.

File: src/pilot/config_idrac.py [hunk 15]
```diff
@@ -322,30 +324,57 @@ def config_idrac(instack_lock,
...
+        target_boot_mode = boot_mode_helper.DRAC_BOOT_MODE_UEFI
+    else:
+        # Set target_boot_mode to BIOS boot mode
```
Comment (rpioso) [inline src/pilot/config_idrac.py]: Again, this comment is unnecessary, because the code is self-explanatory. Please delete it.

File: src/pilot/config_idrac.py [hunk 16]
```diff
@@ -322,30 +324,57 @@ def config_idrac(instack_lock,
...
+        job_ids.append(boot_mode_job_id)
+        if target_boot_mode == boot_mode_helper.DRAC_BOOT_MODE_UEFI:
+            LOG.info("Rebooting {} to apply configuration".format(
+                ip_service_tag))
+            if reboot_required:
+                job_id = drac_client.create_reboot_job()
+                job_ids.append(job_id)
+            drac_client.schedule_job_execution(job_ids,
+                                               start_time='TIME_NOW')
+
```
Comment (rpioso) [inline src/pilot/config_idrac.py]: This blank line contains whitespace. Please keep the blank line, but delete the whitespace. My editor reports the following error:

    W293 blank line contains whitespace

File: src/pilot/config_idrac.py [hunk 17]
```diff
@@ -322,30 +324,57 @@ def config_idrac(instack_lock,
...
+            if reboot_required:
+                job_id = drac_client.create_reboot_job()
+                job_ids.append(job_id)
+            drac_client.schedule_job_execution(job_ids,
+                                               start_time='TIME_NOW')
+    
+            LOG.info("Waiting for iDRAC configuration to \
+                     complete on {}".format(ip_service_tag))
+            LOG.info("Do not unplug {}".format(ip_service_tag))
+
```
Comment (rpioso) [inline src/pilot/config_idrac.py]: This blank line contains whitespace. Please keep the blank line, but delete the whitespace. My editor reports the following error:

    W293 blank line contains whitespace

File: src/pilot/config_idrac.py [hunk 18]
```diff
@@ -302,7 +303,8 @@ def config_idrac(instack_lock,
     drac_user = node["pm_user"]
     drac_password = node["pm_password"]
     ironic_driver = node["pm_type"]
-
```
Comment (rpioso) [inline src/pilot/config_idrac.py]: Restore this blank line.

File: src/pilot/config_idrac.py [hunk 19]
```diff
@@ -362,7 +391,6 @@ def config_idrac(instack_lock,
         job_id = drac_client.create_reboot_job()
         job_ids.append(job_id)
 
-    success = True
```
Comment (rpioso) [inline src/pilot/config_idrac.py]: Restore this line.

File: src/pilot/config_idrac.py [hunk 20]
```diff
@@ -302,7 +303,8 @@ def config_idrac(instack_lock,
     drac_user = node["pm_user"]
     drac_password = node["pm_password"]
     ironic_driver = node["pm_type"]
-
+    job_ids = []
+    success = True
```
Comment (rpioso) [inline src/pilot/config_idrac.py]: Delete this assignment statement, because `success` is not used until the `if` statement that begins on line 394. The definition that was deleted before that statement should be restored. I have added a comment to that effect.

File: src/pilot/config_idrac.py [hunk 21]
```diff
@@ -322,30 +324,57 @@ def config_idrac(instack_lock,
...
+    reboot_required = False
+
+    if BootModeHelper.is_boot_order_flexibly_programmable(drac_client):
+        # Set target_boot_mode to UEFI boot mode
+        target_boot_mode = boot_mode_helper.DRAC_BOOT_MODE_UEFI
+    else:
+        # Set target_boot_mode to BIOS boot mode
+        target_boot_mode = boot_mode_helper.DRAC_BOOT_MODE_BIOS
+
+    job_ids = list()
```
Comment (rpioso) [inline src/pilot/config_idrac.py]: Add a blank line after this line.

File: src/pilot/config_idrac.py [hunk 22]
```diff
@@ -322,30 +324,57 @@ def config_idrac(instack_lock,
     # Clear out any pending jobs in the job queue and fix the condition where
     # there are no pending jobs, but the iDRAC thinks there are
     clear_job_queue(drac_client, ip_service_tag)
+    reboot_required = False
```
Comment (rpioso) [inline src/pilot/config_idrac.py]: Move this assignment statement closer to where `reboot_required` is first used, just after line 336.

    job_ids = list()
    reboot_required = False

    reboot_required_boot_mode, boot_mode_job_id = config_boot_mode(
        drac_client, ip_service_tag, node, target_boot_mode

File: src/pilot/config_idrac.py [hunk 23]
```diff
@@ -322,30 +324,57 @@ def config_idrac(instack_lock,
...
+
+    reboot_required = reboot_required or reboot_required_boot_mode
+    if boot_mode_job_id:
+        job_ids.append(boot_mode_job_id)
+        if target_boot_mode == boot_mode_helper.DRAC_BOOT_MODE_UEFI:
+            LOG.info("Rebooting {} to apply configuration".format(
+                ip_service_tag))
+            if reboot_required:
+                job_id = drac_client.create_reboot_job()
+                job_ids.append(job_id)
```
Comment (rpioso) [inline src/pilot/config_idrac.py]: `reboot_required` needs to be reset to `False`. Otherwise, it could cause an unnecessary reboot in the processing that follows below.

File: src/pilot/config_idrac.py [hunk 24]
```diff
@@ -322,30 +324,57 @@ def config_idrac(instack_lock,
...
-    reboot_required = False
-
     # Configure the NIC port to PXE boot or not
-    reboot_required_nic, job_ids, provisioning_mac = \
+    reboot_required_nic, nic_job_ids, provisioning_mac = \
         configure_nics_boot_settings(drac_client,
                                      ip_service_tag,
                                      pxe_nic_fqdd,
                                      node)
-
```
Comment (rpioso) [inline src/pilot/config_idrac.py]: Restore this blank line.

File: src/pilot/config_idrac.py [hunk 25]
```diff
@@ -322,30 +324,57 @@ def config_idrac(instack_lock,
...
+    if nic_job_ids:
+        job_ids.extend(nic_job_ids)
```
Comment (rpioso) [inline src/pilot/config_idrac.py]: For consistency, move this `if` statement below the next statement.

File: src/pilot/config_idrac.py (hunk 22 above)
Comment (digambar15) [reply src/pilot/config_idrac.py]: Done

File: src/pilot/config_idrac.py (hunk 24 above)
Comment (digambar15) [reply src/pilot/config_idrac.py]: ok

File: src/pilot/config_idrac.py (hunk 21 above)
Comment (digambar15) [reply src/pilot/config_idrac.py]: Done

File: src/pilot/config_idrac.py (hunk 11 above)
Comment (digambar15) [reply src/pilot/config_idrac.py]: Done

File: src/pilot/config_idrac.py (hunk 10 above)
Comment (digambar15) [reply src/pilot/config_idrac.py]: Done

File: src/pilot/config_idrac.py (hunk 8 above)
Comment (digambar15) [reply src/pilot/config_idrac.py]: Done

File: src/pilot/config_idrac.py (hunk 12 above)
Comment (digambar15) [reply src/pilot/config_idrac.py]: Done!

**Comment by digambar15** [comment]
As discussed before sometime, I am following below as suggested by you

1. Squash merge PR #2 to your fork's experimental_UEFI-_support branch. You should be able to do that from the approved PR.
2. Create a topic branch in your fork that points to the new commit created by the step above.
3. Create a PR against the dsp-jetpack/JetPack with the change on the topic branch created in the previous step, step 2.
